File: app/src/main/java/com/waygo/data/base/route/DatabaseRouteBase.java
package com.waygo.data.base.route;

import android.net.Uri;
import android.support.annotation.NonNull;

import rx.android.internal.Preconditions;
import rx.functions.Action1;

public abstract class DatabaseRouteBase implements DatabaseRoute {

    @NonNull
    protected final String tableName;

    public DatabaseRouteBase(@NonNull final String tableName) {
        Preconditions.checkNotNull(tableName, "Table Name cannot be null.");

        this.tableName = tableName;
    }

    @NonNull
    @Override
    public String getTableName() {
        return tableName;
    }

    public void notifyChange(@NonNull Uri uri, @NonNull Action1<Uri> notifyChange) {
        Preconditions.checkNotNull(uri, "URI cannot be null.");
        Preconditions.checkNotNull(notifyChange, "Notify Change function cannot be null.");

        notifyChange.call(uri);
    }
}

File: app/src/main/java/com/waygo/data/base/route/DatabaseRouteBase.kt
package com.waygo.data.base.route

import android.net.Uri

import rx.android.internal.Preconditions
import rx.functions.Action1

public abstract class DatabaseRouteBase(protected val table: String) : DatabaseRoute {

    override fun getTableName(): String = table

    override fun notifyChange(uri: Uri, notifyChange: (Uri) -> Unit ) {
        Preconditions.checkNotNull(uri, "URI cannot be null.")
        Preconditions.checkNotNull(notifyChange, "Notify Change function cannot be null.")

        notifyChange(uri)
    }
}
